[2단계 - 출석 다시 구현하기]  멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?

  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
이번에는 Crew가 자신의 기록에대한 List를 가지고 있는 형태가 아닌, Map을 활용한 객체인 출석부(AttendanceBook)를 통해 작동하도록 설계해봤습니다. 1단계에서 가장 신경쓰였던 디미터 법칙을 준수하도록 노력했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### Optional 사용에 대하여
**기록이 없는 경우에도 수정을 할 수 있도록** 구현하고 싶었는데, Optional 사용이 불가피하였습니다. 또한 도메인에서 찾아오는 값이 Optional 값이기 때문에 컨트롤러에서 Optional을 다루는 로직이 남아있게 되었는데, 이런 구현이 괜찮은지 궁금합니다. 도메인에서 어떻게든 해결해야되는 문제일까요? 최대한 도메인에서 해결하려고 했지만 우선 현재 구현이 최선이었습니다.. 해당 기능을 구현할 수 있는 좋은 방법 있다면 생각을 공유받고 싶습니다..!

`+` 출석이 없는 NoAttendanceRecord (?) 객체 등으로 만들어보는 것도 고려해보습니다. 
그러나 1단계 피드백 과정에서 결석을 직접 저장하지 않고 계산하기로 했던 이유인 메모리 관리와 확장 가능성을 고려해봤을 때, 결석이 없는 경우 또 다른 객체로 만드는 것은 결국 다시 똑같아지는 느낌이 들어 배제하였습니다..

### Dto 사용에 대하여
최근 Dto 사용에 대하여 크루들과 이야기할 기회가 있었습니다. Dto는 단순히 여러 값들을 묶어주기 위해 존재하는 구조가 아니라, View에서 Domain에 의존하는 형태를 dto를 통해 해소할 수 있다고 생각합니다. 

저 또한 dto를 꼭 필요한 때에 사용할 수 있도록 주의하고있는데 이번 미션에서 잘 사용한 것인지 궁금합니다!

### 예외 처리와 검증 책임의 위치
도메인 계층(AttendanceHistory나 AttendanceRecord)이 비즈니스 규칙을 검증하고 예외를 발생시키도록 하는 것과, 컨트롤러에서 이를 미리 확인하여 처리하는 방식 사이에서 고민했습니다.

특히 “이미 출석한 경우, 다시 출석할 수 없으며 수정 기능을 이용하도록 안내”하는 요구사항에서, 사용자 경험 측면(닉네임 입력 후 바로 예외 발생 vs. 등교시간까지 입력 후 예외 발생)과 도메인 책임 분리 사이의 균형이 어려웠습니다.

현재 출석을 할 수 있는 요일인지 검증하는 책임을 가지고있는 WoowaDate 객체의 도입으로 해결할 수 있었지만, 만약 예외 처리 위치가 상관 없었다면 WoowaDate 도입 없이도 해결할 수 있었을 것 같습니다. 다시 구현하면서도 초반에는 WoowaDate 포장 객체를 도입하지 않았는데, 예외처리에 대한 위치를 고민하면서 도입하게 되었습니다. WoowaDate의 도입으로 테스트코드나 기존 코드도 상당히 복잡해져서 이러한 리팩터링이 맞는 방향인지도 고민이었습니다.

yuyoungrhee committed:
chore: 기존 코드 삭제 - step2 준비

yuyoungrhee committed:
docs(README): 요구사항 및 구현할 기능 목록 정리

yuyoungrhee committed:
feat: Application 정의

yuyoungrhee committed:
feat(Crew): 크루 닉네임 검증 기능 구현

yuyoungrhee committed:
feat(AttendanceStatus): 출석 상태 구현

출석, 지각, 결석 상태를 표현하는 enum 클래스로 구현
- 각 출석 상태별 지각 상한 시간(limitTime) 설정
- 지각한 시간으로 상태를 판별하는 기능 구현

yuyoungrhee committed:
feat(AttendanceRecord): 하나의 출석 기록을 관리하는 객체 구현

- 출석한 시간을 통해 출석 상태를 계산하여 반환하는 메서드 구현

yuyoungrhee committed:
feat(AttendanceRecord): 기준 시간과 출석 시간을 기준으로 출석 상태를 판별하는 기능 구현

 - 월요일 기준 시각(13:00)과의 시간 차이를 이용해 출석, 지각, 결석 상태 판별

yuyoungrhee committed:
feat(AttendanceRecord): 모든 요일에 대해서 출석 상태 판별이 가능하도록 구현

- 월요일과 다른요일의 시작시간이 다르다는 점을 반영하여 모든 요일에 대해서 출석 상태가 알맞게 반영되도록 구현

yuyoungrhee committed:
feat(EducationTime): 교육 시간을 Enum으로 관리하도록 구현

yuyoungrhee committed:
feat(AttendanceRecord): 시작시간을 EducationTime을 통해 가져오도록 변경

- 현재 출석 LocalDateTime의 dayOfWeek를 통해 시작시간을 가져오도록 변경

yuyoungrhee committed:
feat(EducationDay): 교육일 판단 로직 구현

- 주말 및 공휴일은 교육일이 아님을 판단하는 로직 구현
- 생성자로 공휴일을 주입받도록 함

yuyoungrhee committed:
feat(AttendanceBook): 크루별 출석 기록 저장 및 조회 기능 구현

yuyoungrhee committed:
feat(AttendanceBook): 출석 수정 기능 구현

- 해당 날짜의 기록을 찾는 기능 구현, 해당 날짜와 동일한지 확인하는 기능을 AttendanceRecord에 추가
- 해당 날짜의 기록을 찾아서 그 기록에게 수정을 요청하도록 구현
- AttendanceRecord에 수정 메서드 추가, 출석 시간을 업데이트하도록 함

yuyoungrhee committed:
feat(WarningStatus): 결석 횟수와 지각 횟수에 따라 경고 수준을 판별하는 기능 구현

yuyoungrhee committed:
feat(AttendanceBook): 특정 크루의 경고 등급을 기록에 따라서 반환하는 기능 작성

yuyoungrhee committed:
feat(AttendanceHistory): 개별 크루의 출석 기록을 관리하는 AttendanceHistory 구현

- 기록 추가, 수정, 날짜로 기록 찾기, 결석 횟수 계산, 지각 횟수 계산 기능을 기존 AttendanceBook -> AttendanceHistory 로 옮김

yuyoungrhee committed:
feat(loader): 파일의 데이터를 읽어와서 객체로 변환하는 기능 구현

- 파일 데이터 자체를 그대로 읽어오는 AttendancesLoader 구현
- 읽어온 데이터를 AttendanceBook 객체로 변환하는 AttendanceAssembler 구현

yuyoungrhee committed:
feat(AttendanceAssembler): indent 1로 줄이도록 수정

yuyoungrhee committed:
feat(InputView): 사용자에게 필요한 입력을 받아오는 기능 구현

- 필요한 Util static 메서드 작성

yuyoungrhee committed:
feat(controller): 1번 출석 확인 기능 구현

- 1번 기능에 해당하는 컨트롤러 작성
- 츌력에 필요한 객체 getter 추가
- OutputView 구현
- MainOption Enum 클래스로 사용자의 기능 선택 구현

yuyoungrhee committed:
feat(controller): 재시도 로직 구현, 메서드 분리

yuyoungrhee committed:
refactor: AttendanceAssembler에서 Book 객체를 직접 생성하는 것이 아니라 Map을 반환하도록 변경

- 출석부에서는 크루 이름이 없으면 추가해야하고, 비즈니스로직에서는 예외를 반환해야하는 차이점을 고려해 만들어진 Map을 주입받도록 변경함

yuyoungrhee committed:
feat(Crews): 크루를 관리하는 Crews 객체 구현

- 크루 이름과 객체를 map으로 관리함
- 크루 이름으로 크루를 찾거나, add하는 기능 구현

yuyoungrhee committed:
feat: Crews도 함께 초기화하도록 구현

- AttendanceBook 테스트에서 해당 크루에 대한 AttendanceHistory를 미리 넣어주도록 변경

yuyoungrhee committed:
feat(AttendanceController): crews 객체도 주입받아 활용하도록 변경

yuyoungrhee committed:
refactor(AttendanceHistory): 출석,지각,결석 횟수를 계산하는 메서드를 하나로 통합

- 파라미터로 출석 상태를 받아 해당하는 상태의 개수를 세는 로직으로 변경함
- view에 개수가 사용되므로 public 으로 변경

yuyoungrhee committed:
feat: 3번 크루별 출석 기록 확인 기능 구현

yuyoungrhee committed:
feat: 2번 출석 수정 기능 구현

- 변경 전과 후의 출석 기록을 출력하기 위해 AttendanceRecord에 자신의 기록을 복사할 수 있는 copy 메서드 추가 구현
- 출석부에서 바로 주어진 날짜에 해당하는 기록을 반환할 수 있도록 중간 메서드 추가 구현

yuyoungrhee committed:
feat: 4번 제적 위험자 확인 기능 추가

yuyoungrhee committed:
feat(AttendanceHistory): 이미 출석한 경우 시나리오 구현

- 이미 출석한 경우, 다시 출석할 수 없으 수정 기능을 이용하도록 안내한다

yuyoungrhee committed:
refactor(AttendanceBook): 필요없는 로직 삭제

- 해당 크루가 존재하는지는 Crews에서 검증하는 것으로 변경되었기 때문에 해당 history null 검증은 필요 없어짐

yuyoungrhee committed:
refactor: 예외 메시지 상수화

yuyoungrhee committed:
refactor(EducationTime): 사용하지 않는 필드 삭제

- endTime은 추후 사용될 가능성이 있으므로 삭제하지 않음

yuyoungrhee committed:
refactor(AttendanceRecord): 메서드명 변경

- LocalDateTime을 가져오는 메서드명 변경, LcalDate를 가져오는 메서드 추가하여 부르는 쪽에서 간단하게 가져올 수 있도록 구현

yuyoungrhee committed:
feat(WoowaDate): 등교일이 아니면 출석할 수 없다는 요구사항 구현

- LocalDate를 감싸는 객체를 만들어 해당 날짜가 교육일인지 아닌지 판단하는 로직 구현
- CustomClock을 사용해 현재 날짜를 컨트롤러에 주입받을 수 있도록 구현

yuyoungrhee committed:
refactor(AttendanceReport): 출석 기록에 대한 통계를 담당하는 Report 구현으로 책임 분리 리팩터링

yuyoungrhee committed:
feat(OutputView): 출석 기록이 없는 날짜에 대해서 출력하는 기능 추가 구현

- AttendanceReport에서 출석 기록이 없는 날짜들을 가져오는 기능 구현

yuyoungrhee committed:
feat(OutputView):  지각을 결석으로 간주하여 내림차순하는 기능 추가 구현

yuyoungrhee committed:
feat: 출석 기록이 없을 때 출석 수정을 이용하여 기록을 추가할 수 있도록 구현

yuyoungrhee committed:
fix(AttendanceController):  메인 옵션에서 오류가 났을 때도 다시 동작하도록 수정

yuyoungrhee committed:
refactor(AttendanceReport): stream을 이용해서 인덴트를 줄임

yuyoungrhee committed:
refactor(WarningStatus): 매직넘버 상수화

yebink commented:
### Optional 사용에 대하여
Optional 도 (지금처럼) 필요하면 써야한다고 생각합니다~ 그런데 함수에서 Optional 을 리턴하는 방식은 그렇게 추천하진 않습니다.
해당 함수를 사용할 때마다 Optional 에 대한 처리가 필요해지고, 그만큼 코드 복잡도가 올라가게 됩니다. 혹시 null 은 고려대상에 없었나요?


### Dto 사용에 대하여
Dto에 대해 학습하셨군요! 데이터 변환만을 하도록 잘 구현하셨습니다 👏 
현재 멍구가 2개의 dto 간 필드가 좀 겹치는 것 같은데, 두 개를 따로 구현하신 이유가 있나요?


### 예외 처리와 검증 책임의 위치
검증 위치는 정말 케바케이고 현업에서도 여전히 고민하는 문제인데요~
개인적으로는 `비즈니스 요구사항에 대한 검증` (ex. 이미 출석한 경우 출석 수정 기능 안내)은 도메인에 넣고, `입력값`에 대한 검증 (ex. 닉네임 empty 체크) 처럼 간단한 것은 요청 객체 에서 하는 것을 선호합니다. 
(나중에 학습하시겠지만, Spring Boot Validation 에서도 입력값 검증에 대한 여러 애너테이션을 지원합니다.)
지금 구현하신 방식도 충분히 좋다고 생각됩니다 😄 

예외 처리를 담당할 곳이 필요해서 WoowaDate 객체를 만드셨군요! 그러다보니 관련한 모든 날짜를 WoowaDate 로 감싸는 행위가 필요했고, 그에 따라 복잡함이 더해졌다고 이해했습니다. 해당 방법도 좋으나, 말씀하신대로 수반되는 불편함이 상당합니다. WoowaDate 내부에 존재하는 예외처리 로직은 `공휴일 또는 주말이 아닌지` 인데, 해당 부분은 이미 EducationDayPolicy 에서 관리되고 있기 때문에, EducationDayPolicy에서 예외처리를 해도 충분하다고 봅니다.
또, EducationDayPolicy는 현재 요구사항에서 변하지 않는 정보이다보니 애플리케이션 내에서 1개만 생성해도 되는데, 지금과 같은 구조는 WoowaDate 마다 새로 생성하게 됩니다.

해당 부분을 고려해 코드를 수정해봐요!

+) 실패하는 테스트 수정 후 다시 요청주시면 꼼꼼히 리뷰해드리겠습니다 😄

yebink reviewed (changes requested):
안녕하세요 멍구~

2단계 잘 구현해주셨네요!
코멘트 확인 후 다시 요청주세요~

yuyoungrhee committed:
test(AttendanceHistory): 실패하는 테스트 코드 수정

- add할때 이미 기록이 있는 날짜인지 확인한 후 add하기 때문에 modify 과정에서도 예외가 발생하는 점을 수정함. modify는 따로 add를 해주도록 수정

yuyoungrhee commented:
### Optional 사용에 대하여
null을 반환하는 것도 고려 대상에 있었으나 호출하는 쪽, 즉 컨트롤러에서 null 검사를 따로 해야한다는 점에서 Optional을 활용한 복잡도와 비교해도 장점이 없다고 느껴졌습니다. 따라서 NPE로부터 자유롭고 명시적으로 해당 변수가 null일 가능성을 표현할 수 있는 Optional을 사용했습니다.
(더 고민해보겠습니다..!)

### Dto 사용에 대하여
> 2개의 dto 간 필드가 좀 겹치는 것 같은데, 두 개를 따로 구현하신 이유가 있나요?

두 DTO가 일부 필드가 겹치더라도 각각 다른 응답을 나타내기 위해 별도로 구현했습니다.

CrewHistoryDto는 크루의 전체 출석 이력, 상세 기록, 누락된 날짜 등 전체적인 이력을 보여주는 3번(출석 확인) 기능에 대한 포맷입니다.
WarningResultDto는 경고 대상자에 대한 요약 정보(결석/지각 횟수, 경고 상태, 유효 결석 횟수)를 전달하는 4번(제적 위험자 확인) 기능에 대한 포맷입니다.
- 만약 각 기능의 View에 대한 요구사항 변경이 있다면, 해당하는 Dto나 View만 변경할 수 있기 때문에 2개로 나눠서 구현했습니다.

### EducationPolicy의 중복 생성에 대하여
EducationPolicy는 현재 Application에서 생성되어 전역에서 동일한 인스턴스를 주입해주고 있습니다. Policy를 싱글톤으로 관리하는 게 좋을지 고민했지만 현재는 Application에서 생성해서 단일 인스턴스를 유지하는 방법으로 구현해도 괜찮다고 생각합니다. 생성 방식에 대해서 제가 잘못 알고 있는 거라면 다시 코멘트 남겨주세요🙂

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,135 @@
+        this.policy = policy;
+    }
+
+    public void run() {
```
1단계 때보다 훨씬 깔끔해졌네요 👏

yebink commented on src/main/java/attendance/controller/MainOption.java:
```diff
@@ -0,0 +1,30 @@
+                .orElseThrow(() -> new IllegalArgumentException(NO_MAIN_OPTION));
+    }
+
+    public boolean continueMain() {
```
미사용으로 보입니다~

yuyoungrhee replied:
삭제했습니다.!

yebink commented on src/main/java/attendance/Application.java:
```diff
@@ -1,47 +1,36 @@
-        CrewRegistry crewRegistry = new CrewRegistry(crews);
-        loadAttendance(attendanceLoader, fixedCustomClock, crewRegistry);
+        CustomClock clock = new CustomClock(LocalDateTime.of(2024, 12, 15, 13, 0));
+        EducationDayPolicy policy = new EducationDayPolicy(Set.of(LocalDate.of(2024, 12, 25)));
```
policy 생성 로직 확인했습니다 👍

yebink commented on src/test/java/attendance/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,79 @@
+        // then
+        AttendanceHistory history = attendanceBook.getHistoryByName(crewName);
+        assertThat(history.getRecords()).hasSize(1);
+        assertThat(history.getRecords().contains(record)).isTrue();
```
```suggestion
        assertThat(history.getRecords()).contains(record);
```

assertj 에서 제공하는 `contains()` 를 쓰면 더 간결하게 표현할 수 있습니다.

yuyoungrhee replied:
그렇군요, 수정했습니다..! 배워갑니다.

yebink commented on src/main/java/attendance/util/DateFormatUtil.java:
```diff
@@ -1,31 +1,22 @@
 import java.time.format.DateTimeParseException;
 
 public class DateFormatUtil {
+    public final static DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("MM월 dd일 EEEE");
```
modifier 순서를 지켜주세요~

```suggestion
    public static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("MM월 dd일 EEEE");
```

https://google.github.io/styleguide/javaguide.html#s4.8.7-modifiers

yebink replied:
해당 내용 전반적으로 체크해주세요!

yuyoungrhee replied:
수정했습니다..! 한 번 잘못사용하니 자각 없이 계속 잘못되게 선언했었군요.. 그 밖의 선언자 순서에 대해서도 다시 한 번 공부할 수 있었습니다. 감사합니다!

yebink commented on src/main/java/attendance/loader/AttendanceAssembler.java:
```diff
@@ -0,0 +1,44 @@
+
+    public Crews assembleCrews(Map<String, AttendanceHistory> histories) {
+        Crews crews = new Crews();
+        histories.keySet().forEach(crewName -> crews.addCrew(new Crew(crewName)));
```
메서드 체이닝은 줄바꿈을 통해 가독성을 높일 수 있습니다~

```suggestion
        histories.keySet()
            .forEach(crewName -> crews.addCrew(new Crew(crewName)));
```

yuyoungrhee replied:
해당 로직은 삭제되어 반영하지 못했지만 앞으로 메서드 체이닝 시 유의하겠습니다!

yebink commented on src/main/java/attendance/domain/Crews.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Crews {
```
Crew의 일급 컬렉션으로 보이네요.

내부적으로 crewMap 이 final 로 선언되어 있지만, addCrew 를 통해 얼마든지 변경이 가능한 구조입니다. 
생성자에서부터 불변임을 보장할 수는 없을까요? 🤔

yuyoungrhee replied:
처음에는 크루 목록도 변할 수 있는 게 당연하다고 생각했습니다. 크루가 추가된다면 addCrew가 필요할테니까요.. 그러나 지금 요구사항에서는 필요없는 로직이었다고 인정합니다.

### 변경사항
생성자에서부터 불변임을 보장할 수 있도록 변경했습니다.
[refactor(Crews): Crews의 불변을 보장하도록 수정](https://github.com/woowacourse/java-attendance/commit/99c7a4edc7067bbe219cb548ed53747f76379bc3)

현재 요구사항에서는 크루를 더 추가할 필요가 없기 때문에 파일데이터로부터 읽어온 크루들이 그대로 Crews에 사용되는 형태입니다. 

초반 파일 데이터 읽어오기 -> 객체 생성 과정에서 이미 정해진 Crews를 생성자 주입 받고, 불변을 보장하도록 하여 변경이 가능합니다.

yebink commented on src/main/java/attendance/Application.java:
```diff
@@ -1,47 +1,36 @@
-        );
+        AttendanceBook attendanceBook = new AttendanceBook(assembleDatas);
+        AttendanceController controller = new AttendanceController(inputView, outputView, attendanceBook, crews, clock,
+                policy);
```
AttendanceAssembler 의 역할은 무엇인가요?

처음엔 csv 파일의 내부 구조 (`Map<String, AttendanceHistory>` 형태) 를 숨기려는 의도로 추측했는데, 26라인에서 다시 꺼내 쓰고 있네요 😅 

해당 클래스가 꼭 필요할까요?

yebink replied:
AttendanceBook 내부에서  `Map<String, AttendanceHistory> crewHistories` 가 아니라`Map<Crew, AttendanceHistory> crewHistories` 를 가진다면 crews 를 꺼내는 과정이 불필요하지 않을까요~?

yuyoungrhee replied:
### AttendanceBook 필드 수정
Map<Crew, AttendanceHistory>로 수정하는 것이 좋을 것 같아 수정했습니다.
[refactor(AttendanceBook): Crew를 key로 사용하도록 수정](https://github.com/YuyoungRhee/java-attendance/commit/df13860e7782847090a38384a01c4f1c0bad170d)

크루와 그에 대한 History를 key-value로 묶는 과정에서 초기에 String으로 키를 설계하여 간단하게 로직을 구현하였는데, 리팩터링 과정에서 수정했으면 좋았을 것 같습니다. 도메인에서 크루는 문자열 이상의 의미를 가지고, AttendanceBook 내에서도 Crew와 History의 관계를 더 명확히 표현할 수 있는 좋은 방법이라고 생각합니다. 더불어 코드도 더 간단해지고요!

### AttendanceAssembler의 역할
[refactor(AttendanceAssembler): AttendanceBook 객체를 완전히 구성하여 반환하도록 수정](https://github.com/woowacourse/java-attendance/commit/d7abeb661f1f464a779c7c9a51a2c21f4c0f7a1d)

Assembler의 역할은 단순 파일 데이터인 크루 이름(String), 출석 시간(LocalDateTIme)을 도메인 객체로 생성하고 조립하는 역할을 합니다. 그 과정에서 최종적으로 Map을 반환하게 되었는데, 다시 생각해보니 **AttendanceBook 자체를 구성해주는 것**이 더 자연스럽다고 느꼈습니다. 아마 Map에서 String을 사용하는 문제 때문에 기존 로직에서는 보지 못했던 문제였던 것 같습니다.(Crews를 다시 Map을 이용해서 생성해줘야하는 문제 때문에요.) Crew를 AttendanceBook의 key로 사용하도록 리팩터링 한 후 고칠 수 있는 방법이 보였습니다.

yebink commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,135 @@
+            WoowaDate targetDate = createTargetDate();
+
+            Optional<AttendanceRecord> findOldRecord = attendanceBook.findRecordBy(crew.getName(), targetDate);
+            AttendanceRecord oldRecord = findOldRecord.map(AttendanceRecord::copy).orElse(null);
```
`copy` 를 하는 이유는 무엇인가요~?

yuyoungrhee replied:
현재 로직에서는 find해서 찾아온 AttendanceRecord 객체의 필드를 수정하기 때문에 같은 객체를 참조하는 findOldRecord에서도 수정이 발생하게 됩니다. 과거 기록의 스냅샷 기능이 필요하여 copy를 사용하게 되었습니다.

copy를 사용하지 않을 시 다음과 같이 변경 전 기록과 변경 후 기록이 같아지게 됩니다.
<img width="856" alt="스크린샷 2025-03-01 오전 2 17 43" src="https://github.com/user-attachments/assets/2c3a2161-2508-415f-9309-069e673eb82b" />

yebink commented on src/main/java/attendance/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,45 @@
+
+public class AttendanceRecord {
+    private final WoowaDate attendanceDate;
+    private LocalTime attendanceTime;
```
final 이 없는 건 의도하신 걸까요~?

yuyoungrhee replied:
우선 의도한 게 맞습니다.
출석 기록은 변할 수 있는 상태로 보고 수정된 객체를 새로 생성하는 것보다 해당 객체를 modify 메서드를 통해 수정하는 것으로 설계했습니다.

attendanceTime도 불변으로 만든다면 modify된 객체를 만드는 방식으로 진행하게 되고 코멘트주신 Controller의 Record `copy`도 사용할 필요 없이 과거의 상태를 보관할 수 있겠습니다..

저는 view에서 과거의 상태를 출력해주는 로직을 제외하고는 굳이 과거 상태를 짧지만 보관(?)할 필요가 없다고 생각했는데, 당일 네오의 수업에서 GC와 JVM 최적화에 대해서 배우면서 요즘은 불변 객체를 새로 만드는 것도 큰 부담이 아닐 수 있을 것 같다는 생각을 하게 됐습니다. 또 copy를 사용하는 것도 결국 copy된 객체가 3번(출석 수정)기능 동안 보관하게되는 거니까요... 비용적인 부분에서는 큰 차이가 나지 않는 걸까요? 

엘리는 AttendanceRecord가 불변을 보장하는 식으로 설계해야한다고 생각하시나요? 안정성을 위해서인가요?

yebink replied:
일단 비용적인 부분에서는, 두 방식 모두 크게 차이가 없다고 봅니다!

해당 코멘트를 드렸던 것은 여러가지 방법 중 copy & modify 방식을 택한 이유가 궁금했기 때문이었고, 제가 궁금했던 부분을 잘 설명해주셨습니다ㅎㅎ 👏 🙂

yebink commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,28 @@
+    ABSENT("결석", Integer.MAX_VALUE);
+
+    private final String title;
+    private final long limitTime;
```
`limitMinute` 과 같이 구체적인 단위를 명시해주세요.

yuyoungrhee replied:
수정했습니다! 고민했던 부분인데 역시 지각한 단위가 분이기 때문에 단위명을 명시해주는 게 좋을 것 같습니다.

yebink commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,28 @@
+public enum AttendanceStatus {
+    PRESENT("출석", 5),
+    LATE("지각", 30),
+    ABSENT("결석", Integer.MAX_VALUE);
```
**특정 날짜에 출석 기록이 없으면 결석** 이 요구사항인데, Integer.MAX_VALUE 와 같은 특정한 값을 선언해버리면 의도치 않은 로직 오류가 발생할 수 있습니다.

`AttendanceStatus.from(Integer.MAX_VALUE)` 의 결과는 ABSENT 입니다.
만약 `일자별 크루들의 지각한 시간(분단위)` 을 저장하는 곳이 있고, 그쪽에서는 반대로 `지각하지 않음`을 표현하기 위해 Integer.MAX_VALUE 를 저장하고 있다면, 해당 크루는 AttendanceStatus 상에서는 ABSENT 인 셈입니다.

이런 식으로 개발자만이 아는 특정 임계값을 사용하는 것은 권장되지 않습니다.

ABSENT 의 경우, 요구사항에 명시된 limitTime 이 없으므로 null 로 명시적으로 해당 부분을 표현해주는 것이 더 좋아보입니다 🙂

yuyoungrhee replied:
null을 임계값으로 설정하는 것에 대해 의문점이 있어 쓰지 못했습니다. 

1. 변수명이 지각 한계값(지각으로 처리되는 마지막 값이라는 의미)인데 값이 없는 null로 설정해도 될까? 

2.  null은 비교 대상이 없으므로 어차피 비즈니스 로직(`lateMiniutes`와 비교해서 해당하는 상태 반환)에서 쓰이지 못하는데 변수로 설정되는 의미가 있는가?

그래서 혹시 이런 방법은 어떨까요? 아니라면 null로 설정해주어도 괜찮은 이유를 설명해주시면 감사하겠습니다..

```java
    private static final int PRESENT_LIMIT_TIME = 5;
    private static final int LATE_LIMIT_TIME = 30;

    ...생략

    public static AttendanceStatus from(long lateMinutes) {
        if (lateMinutes <= PRESENT_LIMIT_TIME) {
            return PRESENT;
        }
        if (lateMinutes <= LATE_LIMIT_TIME) {
            return LATE;
        }
        return ABSENT;
    }
```

---

해당 코멘트에 대한 내용과는 별개로 코멘트에 대해 생각하면서 새롭게 고민하게 된 사항이 있습니다...

결석은 다음과 같습니다.
1. 특정 날짜에 출석 기록이 없으면 결석
2. 지각한 시간이 30분이 초과되면 결석

이 2가지를 분리해서 Status에 저장해야할까요?

제 생각에는 현재 요구사항에서는 같은 결석 처리로 보기 때문에 분리할 필요가 없다고 느낍니다.. 
- 다르게 처리하게 되면 view에서의 --:-- 출력을 Optional 등이 아니라 더 간단하게 처리해볼 수 있을 것 같습니다(구현방법이 정확하게는 그려지지않아 확실치는 않습니다)
- 애초에 출석 기록이 없는 결석 객체를 저장하는 행위에 대해 필요없음을 느끼고 **출석 기록이 없는 결석 시 아예 객체 상태로 보관하지 않았는데** Status를 분리한다면 출석 기록이 없는 객체도 필요한 도메인 로직으로 보고 구현해야할 것 같습니다. 

혹시 다른 의견 있으시면 듣고 싶습니다!

yebink replied:
말씀주신 상수로 설정하는 방법도 괜찮습니다~!
다만 enum에 값이 추가될 때마다 해당 상수도 추가가 필요할지 함께 고려해야 하기 때문에, 확장성 측면에서 아쉬움이 있겠네요 ㅠ

지각 한계값 변수에 `Integer.MAX_VALUE` 라고 설정하신 것은 아마 무한대를 표현하고 싶으셨던 것으로 추측되는데, `Integer.MAX_VALUE` 는 무한대 값이 아닙니다. 마찬가지로 **해당 값만큼 지각할 경우, 결석** 이라는 뜻에 가깝죠 😄 

null 로 설정해주어도 괜찮은 이유는, 지각 한계값 변수가 `출석/지각` 과는 관련이 있지만 `결석` 과는 관련이 없기 때문입니다. 관련이 없다는 것을 명시적으로 null 로 표현해주는 방향으로 의견 드린 것이었습니다 ㅎㅎ

정 null을 피하고 싶다면, 아래와 같이 Predicate 를 사용하는 방법도 있습니다 😄 

```java
public enum AttendanceStatus {
    PRESENT("출석", lateMinutes -> lateMinutes <= 5),
    LATE("지각", lateMinutes -> lateMinutes > 5 && lateMinutes <= 30),
    ABSENT("결석", lateMinutes -> lateMinutes > 30);

    private final String title;
    private final Predicate<Long> isLate;

    AttendanceStatus(String title, Predicate<Long> isLate) {
        this.title = title;
        this.isLate = isLate;
    }
    // 이하생략
```


> 이 2가지를 분리해서 Status에 저장해야할까요?

저도 해당 부분을 분리함으로서 가져갈 수 있는 이점은 잘 모르겠습니다. 이 부분은 유지해도 될 것 같아요!

yuyoungrhee replied:
감사합니다! 충분히 납득 됐어요. 다만 null 사용은 명시적이고 간단한 방법이지만 from을 만드는 과정에서 null을 한 번 검사해주거나 아니면 if문으로 모든 분기를 처리해야한다는 문제가 있었습니다. 따라서 말씀해주신 Predicate를 사용해서 한 번 변경해봤습니다!
[refactor(AttendanceStatus): 조건 로직을 Predicate로 변경](https://github.com/woowacourse/java-attendance/pull/94/commits/92b02831e6137ab75cf921bccbd12561d968c808)

null 사용 시..

```java
public static AttendanceStatus from(long lateMinutes) {
    return Arrays.stream(AttendanceStatus.values())
            .filter(status -> status.limitMinutes != null && lateMinutes <= status.limitMinutes)
            .findFirst()
            .orElse(ABSENT);
}
```

yebink reviewed (changes requested):
안녕하세요 멍구!

전반적으로 잘 구현해주셨습니다 💯 
피드백 확인 후 다시 요청주세요~

yuyoungrhee committed:
refactor(AttendanceBook): Crew를 key로 사용하도록 수정

- 기존 String(크루이름 문자열)에서 Crew 객체 자체를 key로 사용하도록 수정: 현재 크루는 도메인 객체로 문자열 이상의 의미를 가지고 있고, 출석부에서도 이를 반영하여 크루 자체를 key로 사용하는 것이 타당하다고 판단

yuyoungrhee committed:
refactor(AttendanceAssembler): AttendanceBook 객체를 완전히 구성하여 반환하도록 수정

- Assembler는 단순 파일 데이터를 도메인 객체로 구성하는 역할을 하기 때문에, Map을 반환하지 않고 AttendanceBook을 반환하도록 수정함: 상위 계층에서는 내부 구조를 신경쓸 필요가 없도록 캡슐화를 강화

yuyoungrhee committed:
refactor(Crews): Crews의 불변을 보장하도록 수정

- crewMap(현재 crewRegistry)를 외부에서 생성자를 통해 주입받도록 수정
- addCrew 메서드를 삭제하여 외부에서 필드를 추가할 수 없도록 보호함

yuyoungrhee committed:
refactor(MainOption): 미사용 로직 삭제

yuyoungrhee committed:
refactor: 선언자 순서를 지키도록 수정

- static이 final보다 앞에 오도록 수정함

yuyoungrhee committed:
test: 불필요한 contains 호출 제거 - assertj 활용

yuyoungrhee committed:
fix(AttendanceHistory): 출석 수정 기능 오류 수정

- 출석이 존재하지 않을 때만 기록을 새로 추가해야하는데, 추가하는 로직이 if문에서 빠져있어 기록이 존재하는 경우 중복으로 기록이 저장되는 문제 발생
- ifPresentOrElse를 통해 로직 수정

yuyoungrhee committed:
refactor(AttendanceStatus): 변수명을 time에서 minutes로 구체적으로 수정

- 지각한 분을 나타내는 변수명을 구체적인 단위를 명시하도록 수정함

yuyoungrhee commented:
엘리 안녕하세요! 꼼꼼한 코멘트 덕분에 많은 고민을 할 수 있었습니다.

반영한 변경사항은 다음과 같습니다.
1. AttendanceBook 객체의 필드 `Map<String, AttendanceHistory>` -> `Map<Crew, AttendanceHistory>`
2. AttendanceAssembler가 역할을 온전히 할 수 있도록 수정
    - AttendanceBook과 Crews를 assembler 안에서 모두 생성하여 반환해주는 것으로 변경
3. Crews 불변 보장
    - 2번 수정과 더불어 assembelr에서 생성된 crews를 그대로 비즈니스 로직에 사용함으로써 수정 가능했음

각 코멘트에 대한 답변도 달아두었습니다!

yebink reviewed (changes requested):
안녕하세요 멍구~

피드백 잘 반영해주셨습니다 💯 👍 
추가로 피드백 하나 드렸으니 확인해주세요~!

yuyoungrhee committed:
refactor(AttendanceStatus): 조건 로직을 Predicate로 변경

- 변경 이유: 결석의 경우 limitMinutes가 존재하지 않기 때문에 Integer.MAX_VALUE를 지정해주는 것은 올바르지 않은 로직임

yebink reviewed (approved):
안녕하세요 멍구~

이번 미션 고생 많으셨습니다~!
머지하겠습니다 😄
